File: r/R/generate_markdown_report.R
################################################################################
# Script Name: Generate Markdown Reports for CSV Files
# Author: Joao Pedro Azevedo
# Date: 20240409 (original); 2026-05-24 (cso-toolkit hygiene pass)
# Version: 1.1
#
# Description:
# This script processes all CSV files in a specified folder, generates summary
# statistics and descriptive statistics reports in Markdown format for each file.
# The reports include general preamble information, variable details, and optional
# summaries by country, year, and indicator if the necessary columns are present.
#
# Usage:
# - Set the folder path containing CSV files and the output path for saving Markdown reports.
# - Define the column names for country, year, indicator, and value.
# - Call the `process_all_csv_files` function with the specified parameters.
#
# Dependencies:
# - R packages: dplyr, readr, rlang (all calls are namespace-qualified; the
#   helper does NOT attach packages at source-time).
#
# Notes:
# - The script handles missing columns gracefully by skipping summaries if columns
#   are not found and still outputs the preambles.
# - Ensure that the folder paths and column names are correctly specified.
#
# Example:
# folder_path <- "path/to/csv/folder"
# output_path <- "path/to/output/directory"
# process_all_csv_files(folder_path, "countrycode", "year", "indicator", "value", output_path)
#
################################################################################

# NOTE: `.cso_require()` lives in zzz.R as a single shared helper.
# Dependency checks happen INSIDE the public function body
# (generate_markdown_report) below, not at file source-time, so
# vendoring this file does not force the dependency at source().

#' Generate a descriptive-statistics Markdown report from a single CSV file
#'
#' Reads the CSV at `csv_file_path` and writes a Markdown report that
#' includes:
#' \itemize{
#'   \item A general preamble (filename, timestamp, user, number of
#'         observations, number of unique countries / years / indicators,
#'         number of variables).
#'   \item A variable-details table (type, unique cases, mean / SD / min /
#'         max for numerics).
#'   \item Optional summary tables by country, year, and indicator (only
#'         when those columns are present in the input).
#' }
#'
#' Columns that are not present are skipped silently — the preamble still
#' renders.
#'
#' @param csv_file_path Character. Path to the input CSV.
#' @param country_column Character. Column name holding country identifiers.
#' @param year_column Character. Column name holding year values.
#' @param indicator_column Character. Column name holding indicator codes.
#' @param value_column Character. Column name holding the numeric value to
#'   summarise.
#' @param output_path Character. Directory to write `<basename>.md` into.
#'   Default `NULL` writes the file in the current working directory.
#' @param verbose Logical or `NULL`. Show high-level progress and result
#'   messages. `NULL` (default) inherits `getOption("dw.verbose", TRUE)`;
#'   set `TRUE`/`FALSE` to override for this call. See [dw_verbosity()].
#' @param debug Logical or `NULL`. Show internal troubleshooting detail
#'   (resolved paths, dims, branch decisions). `NULL` (default) inherits
#'   `getOption("dw.debug", FALSE)`; implies `verbose`. See [dw_verbosity()].
#'
#' @return Invisibly, `NULL`. Side effect: writes a `.md` file.
#'
#' @examples
#' \dontrun{
#' generate_markdown_report(
#'   "input.csv",
#'   country_column   = "countrycode",
#'   year_column      = "year",
#'   indicator_column = "indicator",
#'   value_column     = "value",
#'   output_path      = "reports/"
#' )
#' }
#' @seealso [process_all_csv_files()] to loop over a folder of CSVs.
#' @family reporting
#' @param verbose Logical or `NULL`. Show high-level progress and result
#'   messages. `NULL` (default) inherits `getOption("dw.verbose", TRUE)`;
#'   set `TRUE`/`FALSE` to override for this call. See [dw_verbosity()].
#' @param debug Logical or `NULL`. Show internal troubleshooting detail.
#'   `NULL` (default) inherits `getOption("dw.debug", FALSE)`; implies
#'   `verbose`. See [dw_verbosity()].
#' @export
generate_markdown_report <- function(csv_file_path, country_column, year_column, indicator_column, value_column, output_path = NULL, verbose = NULL, debug = NULL) {

  .cso_require(c("dplyr", "readr", "rlang"), where = "generate_markdown_report")

  vd <- .dw_vd(verbose, debug); v <- vd$v; d <- vd$d
  .dw_msg("generate_markdown_report", "reading ", basename(csv_file_path), v = v)
  .dw_dbg("generate_markdown_report", "csv_file_path=", csv_file_path, " output_path=", if (is.null(output_path)) "<cwd>" else output_path, d = d)

  # Read the CSV file (generic helper accepting an arbitrary path; dw_use
  # is not applicable because the file is not part of the warehouse layout).
  data <- readr::read_csv(csv_file_path, show_col_types = FALSE)  # cso-allow: io-read-csv

  # Calculate general preamble information
  time_date <- Sys.time()
  user <- Sys.info()[["user"]]
  filename <- basename(csv_file_path)
  num_unique_countries <- if (country_column %in% names(data)) dplyr::n_distinct(data[[country_column]]) else NA
  num_unique_years <- if (year_column %in% names(data)) dplyr::n_distinct(data[[year_column]]) else NA
  num_unique_indicators <- if (indicator_column %in% names(data)) dplyr::n_distinct(data[[indicator_column]]) else NA
  num_variables <- ncol(data)
  num_observations <- nrow(data)  # Number of observations (rows) in the dataset
  .dw_dbg("generate_markdown_report", "rows=", num_observations, " cols=", num_variables, " countries=", num_unique_countries, " years=", num_unique_years, " indicators=", num_unique_indicators, d = d)

  # Function to get variable details
  get_variable_details <- function(data) {
    details <- lapply(names(data), function(var_name) {
      var_data <- data[[var_name]]
      var_type <- if (is.numeric(var_data)) "Numeric" else "String"
      num_unique <- dplyr::n_distinct(var_data)

      if (var_type == "Numeric") {
        mean_val <- mean(var_data, na.rm = TRUE)
        sd_val <- stats::sd(var_data, na.rm = TRUE)
        min_val <- min(var_data, na.rm = TRUE)
        max_val <- max(var_data, na.rm = TRUE)

        c(var_name, var_type, num_unique,
          format(mean_val, scientific = FALSE, big.mark = ",", digits = 2),
          format(sd_val, scientific = FALSE, big.mark = ",", digits = 2),
          format(min_val, scientific = FALSE, big.mark = ",", digits = 2),
          format(max_val, scientific = FALSE, big.mark = ",", digits = 2))
      } else {
        c(var_name, var_type, num_unique, "", "", "", "")
      }
    })

    # Convert list to data frame
    details_df <- as.data.frame(do.call(rbind, details))
    names(details_df) <- c("Variable Name", "Type", "Unique Cases", "Mean", "SD", "Min", "Max")
    details_df
  }

  # Get variable details
  variable_details <- get_variable_details(data)

  # Function to summarize data — `N` is the row count in the group (NOT a
  # distinct-value count; previous label `N_Unique` was misleading).
  summarize_data <- function(data, group_var, value_var) {
    if (value_var %in% names(data)) {
      data |>
        dplyr::group_by(!!rlang::sym(group_var)) |>
        dplyr::summarise(
          N    = dplyr::n(),
          Mean = mean(!!rlang::sym(value_var), na.rm = TRUE),
          SD   = stats::sd(!!rlang::sym(value_var), na.rm = TRUE),
          Min  = min(!!rlang::sym(value_var), na.rm = TRUE),
          Max  = max(!!rlang::sym(value_var), na.rm = TRUE),
          .groups = "drop"
        ) |>
        dplyr::mutate(dplyr::across(dplyr::where(is.numeric), ~ format(.x, scientific = FALSE, big.mark = ",", digits = 2)))
    } else {
      warning(sprintf(
        "[cso_toolkit.generate_markdown_report] Column '%s' not found in data. Skipping summary statistics for '%s'.",
        value_var, group_var
      ), call. = FALSE)
      return(NULL)
    }
  }

  # Summarize by Country (if available)
  summary_by_country <- if (country_column %in% names(data)) {
    summarize_data(data, country_column, value_column)
  } else {
    NULL
  }

  # Summarize by Year (if available)
  summary_by_year <- if (year_column %in% names(data)) {
    summarize_data(data, year_column, value_column)
  } else {
    NULL
  }

  # Summarize by Indicator (if available)
  summary_by_indicator <- if (indicator_column %in% names(data)) {
    summarize_data(data, indicator_column, value_column)
  } else {
    NULL
  }

  # Create Markdown content
  markdown_content <- paste0(
    "# Descriptive Statistics Report\n\n",

    "## General Preamble\n\n",
    "- **Filename**: ", filename, "\n",
    "- **Date and Time**: ", format(time_date, "%Y-%m-%d %H:%M:%S"), "\n",
    "- **User**: ", user, "\n",
    "- **Number of Observations**: ", num_observations, "\n",
    "- **Number of Unique Country Names**: ", ifelse(is.na(num_unique_countries), "N/A", num_unique_countries), "\n",
    "- **Number of Unique Years**: ", ifelse(is.na(num_unique_years), "N/A", num_unique_years), "\n",
    "- **Number of Unique Indicators**: ", ifelse(is.na(num_unique_indicators), "N/A", num_unique_indicators), "\n",
    "- **Number of Variables in the Database**: ", num_variables, "\n\n",

    "## Variable Details Preamble\n\n",
    "| Variable Name | Type | Unique Cases | Mean | SD | Min | Max |\n",
    "|---------------|------|--------------|------|----|-----|-----|\n",
    paste(
      apply(variable_details, 1, function(row) paste("|", paste(row, collapse = " | "), "|")),
      collapse = "\n"
    ), "\n\n"
  )

  # Add summaries only if they are available
  if (!is.null(summary_by_country)) {
    markdown_content <- paste0(markdown_content,
                               "## Summary by Country\n\n",
                               "| ", country_column, " | N | Mean | SD | Min | Max |\n",
                               "|----------------|---------|------|------|-----|-----|\n",
                               paste(
                                 apply(summary_by_country, 1, function(row) paste("|", paste(row, collapse = " | "), "|")),
                                 collapse = "\n"
                               ), "\n\n"
    )
  }

  if (!is.null(summary_by_year)) {
    markdown_content <- paste0(markdown_content,
                               "## Summary by Year\n\n",
                               "| ", year_column, " | N | Mean | SD | Min | Max |\n",
                               "|-------------|---------|------|------|-----|-----|\n",
                               paste(
                                 apply(summary_by_year, 1, function(row) paste("|", paste(row, collapse = " | "), "|")),
                                 collapse = "\n"
                               ), "\n\n"
    )
  }

  if (!is.null(summary_by_indicator)) {
    markdown_content <- paste0(markdown_content,
                               "## Summary by Indicator\n\n",
                               "| ", indicator_column, " | N | Mean | SD | Min | Max |\n",
                               "|----------------|---------|------|------|-----|-----|\n",
                               paste(
                                 apply(summary_by_indicator, 1, function(row) paste("|", paste(row, collapse = " | "), "|")),
                                 collapse = "\n"
                               ), "\n"
    )
  }

  # Generate output file name based on CSV file name
  output_file_name <- paste0(tools::file_path_sans_ext(basename(csv_file_path)), ".md")

  # Determine the full output file path
  if (!is.null(output_path)) {
    output_file <- file.path(output_path, output_file_name)
  } else {
    output_file <- output_file_name
  }

  # Save the markdown content to the output file
  writeLines(markdown_content, con = output_file)
  .dw_msg("generate_markdown_report", "report saved to ", output_file, v = v)
}

#' Generate descriptive-statistics Markdown reports for every CSV in a folder
#'
#' Lists every `.csv` file in `folder_path` and calls
#' [generate_markdown_report()] on each. A thin convenience wrapper.
#'
#' @param folder_path Character. Directory containing input CSVs (not
#'   recursed).
#' @param country_column,year_column,indicator_column,value_column
#'   Column names — see [generate_markdown_report()].
#' @param output_path Character. Output directory for `.md` files. Default
#'   `NULL` writes into the current working directory.
#' @param verbose Logical or `NULL`. Show high-level progress and result
#'   messages. `NULL` (default) inherits `getOption("dw.verbose", TRUE)`;
#'   set `TRUE`/`FALSE` to override for this call. See [dw_verbosity()].
#' @param debug Logical or `NULL`. Show internal troubleshooting detail
#'   (resolved paths, dims, branch decisions). `NULL` (default) inherits
#'   `getOption("dw.debug", FALSE)`; implies `verbose`. See [dw_verbosity()].
#'
#' @return Invisibly, `NULL`.
#'
#' @examples
#' \dontrun{
#' process_all_csv_files(
#'   folder_path     = "data/raw/",
#'   country_column   = "countrycode",
#'   year_column      = "year",
#'   indicator_column = "indicator",
#'   value_column     = "value",
#'   output_path      = "reports/"
#' )
#' }
#' @seealso [generate_markdown_report()] (single-file engine).
#' @family reporting
#' @param verbose Logical or `NULL`. Show high-level progress and result
#'   messages. `NULL` (default) inherits `getOption("dw.verbose", TRUE)`;
#'   set `TRUE`/`FALSE` to override for this call. See [dw_verbosity()].
#' @param debug Logical or `NULL`. Show internal troubleshooting detail.
#'   `NULL` (default) inherits `getOption("dw.debug", FALSE)`; implies
#'   `verbose`. See [dw_verbosity()].
#' @export
process_all_csv_files <- function(folder_path, country_column, year_column, indicator_column, value_column, output_path = NULL, verbose = NULL, debug = NULL) {
  # List all CSV files in the folder
  csv_files <- list.files(path = folder_path, pattern = "\\.csv$", full.names = TRUE)

  # Loop through each CSV file and generate a Markdown report
  vd <- .dw_vd(verbose, debug); v <- vd$v; d <- vd$d
  .dw_msg("process_all_csv_files", "found ", length(csv_files), " CSV file(s) in ", folder_path, v = v)
  .dw_dbg("process_all_csv_files", "folder_path=", folder_path, " output_path=", if (is.null(output_path)) "<cwd>" else output_path, d = d)

  for (csv_file in csv_files) {
    .dw_msg("process_all_csv_files", "processing ", csv_file, v = v)
    generate_markdown_report(csv_file, country_column, year_column, indicator_column, value_column, output_path, verbose = v, debug = d)
  }
}

# =============================================================================
# v0.4.5 — dw_-prefixed canonical aliases (issue #42)
# =============================================================================

#' @rdname generate_markdown_report
#' @export
dw_generate_markdown_report <- generate_markdown_report

#' @rdname process_all_csv_files
#' @export
dw_process_all_csv_files <- process_all_csv_files

File: python/src/generate_markdown_report.py
"""Generate descriptive-statistics Markdown reports from CSV files.

Python port of ``r/R/generate_markdown_report.R``.

* :func:`generate_markdown_report` — process a single CSV.
* :func:`process_all_csv_files` — loop over a folder of CSVs.
"""

from __future__ import annotations

import datetime as _dt
import getpass
import os
import warnings
from pathlib import Path
from typing import List, Optional, Union

import numpy as np
import pandas as pd


def _fmt(x: float) -> str:
    """Format a number with comma thousands separators and ~2 sig figs."""
    if pd.isna(x):
        return ""
    if abs(x) >= 1 or x == 0:
        return f"{x:,.2f}"
    return f"{x:.3g}"


def _variable_details(data: pd.DataFrame) -> pd.DataFrame:
    """Build the per-variable summary table."""
    rows: List[List[str]] = []
    for name in data.columns:
        col = data[name]
        if pd.api.types.is_numeric_dtype(col):
            rows.append([
                name, "Numeric", str(col.nunique(dropna=True)),
                _fmt(col.mean()), _fmt(col.std()),
                _fmt(col.min()), _fmt(col.max()),
            ])
        else:
            rows.append([name, "String", str(col.nunique(dropna=True)),
                         "", "", "", ""])
    return pd.DataFrame(
        rows,
        columns=["Variable Name", "Type", "Unique Cases",
                 "Mean", "SD", "Min", "Max"],
    )


def _summarize(data: pd.DataFrame, group_var: str,
               value_var: str) -> Optional[pd.DataFrame]:
    """Summarise ``value_var`` grouped by ``group_var``."""
    if value_var not in data.columns:
        warnings.warn(
            f"Column {value_var!r} not found in data. "
            f"Skipping summary statistics for {group_var}.",
            stacklevel=2,
        )
        return None
    val = pd.to_numeric(data[value_var], errors="coerce")
    grouped = (
        data.assign(_value=val)
        .groupby(group_var, dropna=False)["_value"]
        .agg(N="count", Mean="mean", SD="std", Min="min", Max="max")
        .reset_index()
    )
    for col in ("N", "Mean", "SD", "Min", "Max"):
        grouped[col] = grouped[col].apply(_fmt) if col != "N" else grouped[col].astype(str)
    return grouped


def _md_table(df: pd.DataFrame) -> str:
    """Render a DataFrame as a GitHub-flavored Markdown table."""
    if df is None or len(df) == 0:
        return ""
    header = "| " + " | ".join(df.columns) + " |"
    sep = "|" + "|".join(["---"] * len(df.columns)) + "|"
    rows = ["| " + " | ".join(str(v) for v in row) + " |"
            for row in df.itertuples(index=False, name=None)]
    return "\n".join([header, sep, *rows])


def generate_markdown_report(
    csv_file_path: Union[str, Path],
    country_column: str,
    year_column: str,
    indicator_column: str,
    value_column: str,
    output_path: Optional[Union[str, Path]] = None,
) -> None:
    """Generate a descriptive-statistics Markdown report from one CSV.

    Reads the CSV at ``csv_file_path`` and writes a Markdown report
    containing a general preamble (filename, timestamp, user, number of
    observations, unique-country / -year / -indicator counts, number of
    variables), a variable-details table, and optional summary tables by
    country, year, and indicator (only when the corresponding columns
    are present).

    Parameters
    ----------
    csv_file_path
        Path to the input CSV.
    country_column, year_column, indicator_column
        Column names; absent columns are skipped silently.
    value_column
        Numeric column to summarise.
    output_path
        Directory to write ``<basename>.md`` into.  ``None`` writes to
        the current working directory.
    """
    csv_file_path = Path(csv_file_path)
    data = pd.read_csv(csv_file_path)  # cso-allow: io-read-csv

    time_date = _dt.datetime.now().strftime("%Y-%m-%d %H:%M:%S")
    user = os.environ.get("USERNAME") or getpass.getuser()
    filename = csv_file_path.name

    n_countries = data[country_column].nunique() if country_column in data.columns else None
    n_years = data[year_column].nunique() if year_column in data.columns else None
    n_indicators = data[indicator_column].nunique() if indicator_column in data.columns else None
    n_variables = len(data.columns)
    n_obs = len(data)

    var_details = _variable_details(data)

    summaries: dict[str, Optional[pd.DataFrame]] = {
        "Country": _summarize(data, country_column, value_column) if country_column in data.columns else None,
        "Year": _summarize(data, year_column, value_column) if year_column in data.columns else None,
        "Indicator": _summarize(data, indicator_column, value_column) if indicator_column in data.columns else None,
    }

    lines = [
        "# Descriptive Statistics Report",
        "",
        "## General Preamble",
        "",
        f"- **Filename**: {filename}",
        f"- **Date and Time**: {time_date}",
        f"- **User**: {user}",
        f"- **Number of Observations**: {n_obs}",
        f"- **Number of Unique Country Names**: {n_countries if n_countries is not None else 'N/A'}",
        f"- **Number of Unique Years**: {n_years if n_years is not None else 'N/A'}",
        f"- **Number of Unique Indicators**: {n_indicators if n_indicators is not None else 'N/A'}",
        f"- **Number of Variables in the Database**: {n_variables}",
        "",
        "## Variable Details Preamble",
        "",
        _md_table(var_details),
        "",
    ]

    section_labels = [("Country", country_column),
                      ("Year", year_column),
                      ("Indicator", indicator_column)]
    for label, col in section_labels:
        df = summaries[label]
        if df is None:
            continue
        lines.append(f"## Summary by {label}")
        lines.append("")
        lines.append(_md_table(df))
        lines.append("")

    out_name = csv_file_path.stem + ".md"
    out_file = Path(output_path) / out_name if output_path else Path(out_name)
    out_file.parent.mkdir(parents=True, exist_ok=True)
    out_file.write_text("\n".join(lines), encoding="utf-8")
    print(f"Markdown report saved to {out_file}")


def process_all_csv_files(
    folder_path: Union[str, Path],
    country_column: str,
    year_column: str,
    indicator_column: str,
    value_column: str,
    output_path: Optional[Union[str, Path]] = None,
) -> None:
    """Generate descriptive-statistics reports for every CSV in a folder.

    Lists every ``.csv`` file in ``folder_path`` (non-recursive) and
    calls :func:`generate_markdown_report` on each.

    Parameters
    ----------
    folder_path
        Directory containing input CSVs.
    country_column, year_column, indicator_column, value_column
        Forwarded to :func:`generate_markdown_report`.
    output_path
        Output directory for ``.md`` files.
    """
    folder = Path(folder_path)
    for csv_file in sorted(folder.glob("*.csv")):
        print(f"Processing file: {csv_file}")
        generate_markdown_report(
            csv_file, country_column, year_column,
            indicator_column, value_column, output_path,
        )
